File: e62/java/Solution.java
public class Solution {
    /** 
     *@param A : an integer rotated sorted array
     *@param target :  an integer to be searched
     *return : an integer
     */
    public int search(int[] A, int target) {
        // write your code here
        int size = A.length;
        if (size == 0) {
            return -1;
        }
        int shift = findMin(A);
        //System.out.println(shift);
        int l = 0;
        int r = size - 1;
        if (target == A[0+shift]) {
            return 0+shift;
        }
        if (target == A[(size-1+shift) % size]) {
            return (size-1+shift) % size;
        }
        while (r -l > 1) {
            int mid = (l + r) / 2;
            int realidx = (mid + shift) % size;
            int tmp = A[realidx];
            if (tmp == target) {
                return realidx;
            }
            else if (tmp > target) {
                r = mid;
            }
            else {
                l = mid;
            }
        }
        return -1;
    }
    
    /**
     * @param nums: a rotated sorted array
     * @return: the minimum number in the array
     */
    public static int findMin(int[] nums) {
        // write your code here
        int size = nums.length;
        if (nums[0] < nums[size - 1]) {
            return 0;
        }
        int left = 0;
        int right = size - 1;
        while (right - left > 1) {
            int mid = (left + right) / 2;
            if (nums[mid] > nums[right]) {
                left = mid;
            }
            else {
                right = mid;
            }
        }
        return right;
    }
}

File: e62/python/solution.py
class Solution:
    """
    @param A : a list of integers
    @param target : an integer to be searched
    @return : an integer
    """
    def search(self, A, target):
        # write your code here
        if A is None or len(A) == 0:
            return -1
        l = 0
        r = len(A) - 1
        t = target
        if target == A[0]:
            return 0
        if target == A[len(A)-1]:
            return len(A) - 1
        while r - l > 1:
            m = (l + r) / 2
            if A[l] < A[m] < A[r] < t or \
               A[r] < t < A[l] < A[m] or \
               A[m] < A[r] < t < A[l] or \
               t < A[l] < A[m] < A[r]:
                return -1
            if t == A[m]:
                return m
            elif A[l] < A[m] < t < A[r] or \
                 A[r] < A[l] < A[m] < t or \
                 A[m] < t < A[r] < A[l] or \
                 t < A[r] < A[l] < A[m]:
                l = m
            else:
                r = m
        return -1
                
